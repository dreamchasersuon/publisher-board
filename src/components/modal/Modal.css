.Overlay {
  position: fixed;
  top: 0;
  left: 0;
  display: flex;
  align-content: center;
  align-items: center;
  flex-direction: column;
  justify-content: center;
  height: 100vh;
  width: 100vw;
  overflow: auto;
}

.Overlay-backdrop {
  position: fixed;
  top: 0;
  left: 0;
  bottom: 0;
  width: 100%;
  height: 100%;
  opacity: 0.25;
  background-color: #000;
  z-index: 2000;
  transition: opacity 0.6s cubic-bezier(0.47, 0, 0.745, 0.715);
}

.Modal {
  width: 420px;
  border-radius: 4px;
  background-color: white;
  z-index: 3000;
  box-shadow: 0 0 30px rgba(0, 0, 0, 0.25);
}
