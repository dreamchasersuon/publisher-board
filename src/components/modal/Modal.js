import React, { useRef, useState } from 'react';
import PropTypes from 'prop-types';
import ModalHeader from './header/ModalHeader';
import ModalBody from './body/ModalBody';
import ModalFooter from './footer/ModalFooter';
import FormUpdateUser from '../forms/update-user/FormUpdateUser';
import FormAddUser from '../forms/add-user/FormAddUser';
import FormUpdateBalance from '../forms/update-balance/FormUpdateBalance';
import FormGetTransactions from '../forms/get-transactions/FormGetTransactions';
import Transactions from '../transactions/Transactions';
import './Modal.css';

const forms = {
  updateUser: {
    render(props) {
      return <FormUpdateUser {...props} />;
    },
    btnCloseLabel: 'cancel',
    btnProceedLabel: 'update',
    title: 'Update User',
  },
  addUser: {
    render(props) {
      return <FormAddUser {...props} />;
    },
    btnCloseLabel: 'cancel',
    btnProceedLabel: 'add',
    title: 'Add User',
  },
  updateBalance: {
    render(props) {
      return <FormUpdateBalance {...props} />;
    },
    btnCloseLabel: 'cancel',
    btnProceedLabel: 'update',
    title: 'Update Balance',
  },
  getTransactions: {
    render(props) {
      return <FormGetTransactions {...props} />;
    },
    btnCloseLabel: 'cancel',
    btnProceedLabel: 'get',
    title: 'Get Transactions',
  },
  transactions: {
    render(props) {
      return <Transactions {...props} />;
    },
    btnCloseLabel: 'close',
    btnProceedLabel: null,
    title: 'Transactions',
  },
  get(obj, name) {
    return this.hasOwnProperty(name)
      ? this[name]
      : {
          render() {
            return <span>Nothing to show</span>;
          },
          btnCloseLabel: 'close',
          title: 'Body not found',
        };
  },
};

const formsMap = new Proxy({}, forms);

export default function Modal({ name, onClose }) {
  const [isFetching, setIsFetching] = useState(false);
  const formRef = useRef(null);
  function handleOnSubmitForm() {
    if (formRef.current) {
      formRef.current.handleSubmit();
    }
  }

  return (
    <div className="Overlay">
      <div onClick={onClose} className="Overlay-backdrop" />
      <div className="Modal">
        <ModalHeader title={formsMap[name].title} />
        <ModalBody>
          {formsMap[name].render({ formRef, onClose, setIsFetching })}
        </ModalBody>
        <ModalFooter
          isLoading={isFetching}
          handleOnSubmitForm={handleOnSubmitForm}
          handleOnCloseModal={onClose}
          btnCloseLabel={formsMap[name].btnCloseLabel}
          btnProceedLabel={formsMap[name].btnProceedLabel}
        />
      </div>
    </div>
  );
}

Modal.propTypes = {
  name: PropTypes.string,
  onClose: PropTypes.func,
};
