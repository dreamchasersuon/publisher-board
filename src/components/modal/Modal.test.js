import React from 'react';
import snapshotDiff from 'snapshot-diff';
import Modal from './Modal';

const handleOnPress = jest.fn();

jest.mock('../forms/add-user/FormAddUser', () => 'FormAddUser');

test('renders modal with form selected by name or nothing to show message', () => {
  expect(
    snapshotDiff(
      <Modal onClose={handleOnPress} name="addUser" />,
      <Modal onClose={handleOnPress} />,
    ),
  ).toMatchSnapshot();
});
