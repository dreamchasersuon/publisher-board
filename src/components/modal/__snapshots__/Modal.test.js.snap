// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders modal with form selected by name or nothing to show message 1`] = `
"Snapshot Diff:
- <Modal name=\\"addUser\\" onClose={[Function mockConstructor]} />
+ <Modal onClose={[Function mockConstructor]} />

@@ -10,42 +10,28 @@
    >
      <header
        className=\\"Modal-header\\"
      >
        <h2>
-         Add User
+         Body not found
        </h2>
      </header>
      <main
        className=\\"Modal-body\\"
      >
-       <FormAddUser
-         formRef={
-           Object {
-             \\"current\\": null,
-           }
-         }
-         onClose={[Function mockConstructor]}
-         setIsFetching={[Function bound dispatchAction]}
-       />
+       <span>
+         Nothing to show
+       </span>
      </main>
      <footer
        className=\\"Modal-footer\\"
      >
        <button
          className=\\"Modal-action\\"
          onClick={[Function mockConstructor]}
          type=\\"button\\"
        >
-         cancel
-       </button>
-       <button
-         className=\\"Modal-action\\"
-         data-testid=\\"modal-perform\\"
-         onClick={[Function handleOnSubmitForm]}
-         type=\\"button\\"
-       >
-         add
+         close
        </button>
      </footer>
    </div>
  </div>"
`;
